examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,292 @@
+    dropout = model_cfg.dropout
+
+    batch_size = data_cfg.batch_size
+
+    grad_clip = train_cfg.grad_clip
+    max_epochs = train_cfg.max_epochs
+    always_save_checkpoint = train_cfg.always_save_checkpoint
+
+    episode_length = ppo_cfg.episode_length
+    ppo_batch_size = ppo_cfg.ppo_batch_size
+    ppo_num_epochs = ppo_cfg.ppo_num_epochs
+    num_rollouts_per_epoch = ppo_cfg.num_rollouts_per_epoch
+
+    device = cfg.sys.device
+    dtype = cfg.sys.dtype
+    compile_ = cfg.sys.compile
+
+    ctx = setup(device, dtype)
+
+    train_loader = get_dataloader(
+        data_cfg.batch_size,
+        data_cfg.block_size,
+        PromptData,
+        device,
+        dataset_name="CarperAI/openai_summarize_tldr",
+        split="train",
+    )
+    val_loader = get_dataloader(
+        data_cfg.batch_size,
+        data_cfg.block_size,
+        PromptData,
+        device,
+        dataset_name="CarperAI/openai_summarize_tldr",
+        split="valid",
+    )
+
+    actor, critic, critic_head, model = init_actor_critic(
+        resolve_name_or_path(transformer_name_or_path), dropout, device, compile_
+    )
+    ref_model = deepcopy(model).to("cuda:1")
+    ref_model.requires_grad_(False)
+    layers = model.transformer.h
+    num_layers = len(layers)
+    num_unfrozen = int(0.3 * num_layers)
+    for layer in layers[:-num_unfrozen]:
+        layer.requires_grad_(False)
+
+    reward_model = init_reward_model(
+        reward_model_path=resolve_name_or_path(reward_model_cfg.name_or_path),
+        device=device,
+        compile_=compile_,
+    )
+    reward_model.eval()
+    reward_model.requires_grad_(False)
+
+    adv_fn = GAE(
+        value_network=critic, gamma=0.99, lmbda=0.95, average_gae=True, shifted=True
+    )
+    loss_fn = ClipPPOLoss(actor, critic_head)
+
+    test_prompt = next(val_loader)
+    reward_estimator = RewardEstimator(
+        eval_iters, episode_length, reward_model, ref_model
+    )
+
+    prompt_logger = TestPromptLogger(
+        batch=test_prompt,
+        reward_model=reward_model,
+        logger=query_logger,
+        episode_length=episode_length,
+    )
+
+    optimizer = torch.optim.AdamW(
+        [p for p in loss_fn.parameters() if p.requires_grad], **train_cfg.optimizer
+    )
+    scheduler = None
+    if train_cfg.decay_lr:
+        scheduler = CosineAnnealingLR(optimizer, **train_cfg.scheduler)
+
+    rb = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(episode_length * num_rollouts_per_epoch),
+        batch_size=episode_length * batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+    rb_ppo = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(episode_length * batch_size),
+        batch_size=ppo_batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+
+    rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+    kl_controller = AdaptiveKLController(rollout_from_model, 0.1, 6, 10000)
```

Comment:
I don't think AdaptiveKLController takes the rollout_from_model as input does it?